File: src/main/java/com/zzzj/acw/Q1068.java
package com.zzzj.acw;

import java.util.Scanner;

public class Q1068 {


    public static void main(String[] args) {

        Scanner scanner = new Scanner(System.in);

        int M = scanner.nextInt();

        int N = M << 1;

        int[] stone = new int[N];

        for (int i = 0; i < M; i++) {
            int num = scanner.nextInt();
            stone[i] = num;
            stone[i + M] = num;
        }

        int[][] max = new int[N][N];

        long[][] min = new long[N][N];

        int[] preSum = new int[N + 1];

        for (int i = 1; i <= N; i++)
            preSum[i] = preSum[i - 1] + stone[i - 1];

        for (int len = 1; len < M; len++) {

            for (int left = 0; left < N; left++) {

                int right = left + len;

                if (right >= N)
                    break;

                // System.out.printf("left = %d , right = %d %n", left, right);

                min[left][right] = Integer.MAX_VALUE;

                for (int k = left; k < right; k++) {

                    max[left][right] = Math.max(
                            max[left][right],
                            max[left][k] + max[k + 1][right] + preSum[right + 1] - preSum[left]
                    );

                    min[left][right] = Math.min(
                            min[left][right],
                            min[left][k] + min[k + 1][right] + preSum[right + 1] - preSum[left]
                    );

                }

            }

        }

        int maxV = Integer.MIN_VALUE;

        int minV = Integer.MAX_VALUE;

        for (int i = 0; i < M; i++) {
            maxV = Math.max(
                    maxV,
                    max[i][i + M - 1]
            );
            minV = (int) Math.min(
                    minV,
                    min[i][i + M - 1]
            );
        }

        System.out.println(minV);
        System.out.println(maxV);
    }

}

File: src/main/cpp/acw/Q1068.cpp
#include "iostream"
#include "cstring"
#include "algorithm"

using namespace std;

const int N = 420;

const int INF = 1E8;

int a[N];

int f[N][N];
int f2[N][N];

int s[N];

#include "zzzj.h"

int main() {

#ifdef CPP_ZZZJ_H
    Zzzj::reopen(__FILE__);
#endif

    int n;

    cin >> n;

    for (int i = 0; i < n; ++i) {
        int v;
        cin >> v;
        a[i] = a[i + n] = v;
    }

    int m = n << 1;

    for (int i = 0; i < m; ++i) {
        for (int j = 0; j < m; ++j) {
            f2[i][j] = INF;
        }
        f2[i][i] = 0;
    }

    for (int i = 1; i <= m; ++i)
        s[i] = s[i - 1] + a[i];

    for (int len = 1; len < m; ++len) {

        for (int left = 0; left + len < m; ++left) {

            int right = left + len;

            for (int k = left; k < right; ++k) {
                f[left][right] = max(f[left][right], f[left][k] + f[k + 1][right] + s[right + 1] - s[left]);
                f2[left][right] = min(f2[left][right], f2[left][k] + f2[k + 1][right] + s[right + 1] - s[left]);
            }

        }

    }

    int maxv = 0;
    int minv = INF;

    for (int i = 0; i < n; ++i) {
        maxv = max(maxv, f[i][i + n - 1]);
        minv = min(minv, f2[i][i + n - 1]);
    }

    cout << minv << endl;
    cout << maxv << endl;
}